File: 백준/Gold/1863. 스카이라인 쉬운거/스카이라인 쉬운거.js
const [_, ...input] = require("fs").readFileSync("/dev/stdin").toString().trim().split('\n');
const stack = [0];
let answer = 0;

for (let i of input) {
    const [x, y] = i.split(' ').map(Number);

    if (stack.at(-1) < y) {
        answer++;
        stack.push(y);
    }

    if (stack.at(-1) > y) {
        if (stack.indexOf(y) === -1 && y !== 0) {
            answer++;
        }

        while(stack.at(-1) >= y) {
            stack.pop();
        }

        stack.push(y);
    }
}

console.log(answer);

/*
currHeight보다 높으면 + 1 / set추가 / currHeight 갱신.
이전보다 낮아지면 갱신 / set에 없으면 + 1 / 그 이상의 set들 전부 삭제.
*/ 

File: 백준/Gold/1863. 스카이라인 쉬운거/스카이라인 쉬운거.cc
#include <stdio.h>
#include <stack>
using namespace std;

int main()
{
  stack<int> s;
  int n, answer = 0;
  scanf("%d", &n);

  s.emplace(0);

  while (n--)
  {
    int x, y;
    scanf("%d %d", &x, &y);

    if (!s.empty() && s.top() < y) // top보다 큰 값이 들어오면 asnwer++ & stack 추가.
    {
      answer++;
      s.emplace(y);
      continue;
    }

    if (!s.empty() && s.top() == y)
    {
      continue;
    }

    if (!s.empty() && s.top() > y)
    {
      bool duplicated = false;
      while (!s.empty() && s.top() >= y)
      {
        if (s.top() == y)
        {
          duplicated = true;
        }
        s.pop();
      }

      if (!duplicated)
      {
        answer++;
      }

      s.emplace(y);
    }
  }

  printf("%d", answer);

  return 0;
}